### DISCUSSION TITLE
findFirstRecordByData with a relation

### DISCUSSION THREAD

**Discussion by TheHypnoo** [pocketbase/pocketbase#5595] [Q&A] [answered]
Hi, I am trying to create a custom endpoint, in this case I want to get a record from an id, that id is inside a relation.

That is, I give you a simple example:

zones: {
    sensors: [‘idSensor1’, ‘idSensor2’] -> this is a relation of another table
}

Now, how can I use findFirstRecordByData to get the sensors from a specific id.

**Comment by orenaksakal** [comment]
you are looking for `expand` https://pocketbase.io/docs/api-records/#listsearch-records

**Reply by TheHypnoo** [reply]
No, I mean this section: https://pocketbase.io/docs/js-records/#fetch-single-record

Where you can search for a record by its information, but when it is an array, how can you search for it?

**Reply by ganigeorgiev** [reply, marked as answer]
@TheHypnoo `FindFirstRecordByData` retrieves a single record by the exact field key-value match.

If you want to search within a multiple valued field you'll have to either [construct a custom RecordQuery](https://pocketbase.io/docs/js-records/#custom-record-query) and use the SQLite [`json_each` function](https://www.sqlite.org/json1.html) OR simply use the `findFirstRecordByFilter` helper (it uses the same syntax as the API rules and the client-side Web APIs `filter` query param):

```js
const record = $app.dao().findFirstRecordByFilter(
    "zones",
    "sensors.id ?= {:sensorId}",
    { sensorId: "idSendor1" },
)
```

_If that doesn't help, please be more specific about your collections structure and the desired behavior._

**Reply by TheHypnoo** [reply]
That was just the answer I needed. 

Thank you very much.

**Reply by TheHypnoo** [reply]
Hello! I would like to know just the same way but instead of with one element, get multiple elements. How would it be?

With a Custom Query it should be? how? could you give me an example. Thank you!

@ganigeorgiev

**Comment by dschissler** [comment]
Here are a few examples query using Golang as a Framework.  I put all of these custom queries in a single `models` package along with the constants and `New` functions (where `models` here is the pb models package reference).  I don't define these queries directly in my API controllers in order to reduce the brittleness.

```golang
func SitesCollection(dao *daos.Dao) (*models.Collection, error) {
	return dao.FindCollectionByNameOrId(Sites)
}

func NewSite(dao *daos.Dao) (*models.Record, error) {
	if col, err := SitesCollection(dao); err != nil {
		return nil, err
	} else {
		return models.NewRecord(col), nil
	}
}

func SiteByPointName(dao *daos.Dao, name string) (*models.Record, error) {
	m := &models.Record{}
	err := dao.RecordQuery(Sites).
		LeftJoin(SitesPoints, dbx.NewExp("sites_points.site = sites.id")).
		LeftJoin(Points, dbx.NewExp("points.point = point.id")).
		AndWhere(dbx.HashExp{
			"points.name": name,
			"points.type": PointTypeSite,
		}).
		Limit(1).
		One(m)
	if err != nil {
		return nil, err
	}

	return m, nil
}
```

For multiple records:

```golang
func UsersByMembership(dao *daos.Dao, siteId string, filter MembersFilter) ([]*models.Record, error) {
	q := dao.RecordQuery(Users).
		LeftJoin(SitesUsers, dbx.NewExp("sites_users.user = users.id")).
		AndWhere(dbx.HashExp{"sites_users.site": siteId})

	if filter.Type != "" {
		q.AndWhere(dbx.HashExp{"sites_users.type": filter.Type})
	}

	if filter.TypeAtLeast != "" {
		q.AndWhere(dbx.In("sites_users.type",
			slice.Any(MembershipAtLeast(filter.TypeAtLeast))...,
		))
	}

	records := []*models.Record{}
	err := q.All(&records)
	return records, err
}
```
